libclc/utils/libclc-remangler/LibclcRemangler.cpp

```diff
@@ -147,340 +123,606 @@ class BumpPointerAllocator {
+
+    // When in test run, don't enable replacements and assert that re-mangled
+    // name matches the original.
+    if (!TestRun) {
+      auto It = TypeReplacements.find(Name.begin());
+      if (It != TypeReplacements.end())
+        Name = StringView(It->second);
+    }
 
-    const FunctionEncoding *encoding =
-        static_cast<const FunctionEncoding *>(root);
+    clang::QualType Res;
+    bool IsVariadic = false;
+
+    // First find the match against `QualType`...
+    if (Name == "void")
+      Res = ContextAST->VoidTy;
+    else if (Name == "wchar_t")
+      Res = ContextAST->WCharTy;
+    else if (Name == "bool")
+      Res = ContextAST->BoolTy;
+    else if (Name == "char")
+      Res = ContextAST->CharTy;
+    else if (Name == "signed char")
+      Res = ContextAST->SignedCharTy;
+    else if (Name == "unsigned char")
+      Res = ContextAST->UnsignedCharTy;
+    else if (Name == "short")
+      Res = ContextAST->ShortTy;
+    else if (Name == "unsigned short")
+      Res = ContextAST->UnsignedShortTy;
+    else if (Name == "int")
+      Res = ContextAST->IntTy;
+    else if (Name == "unsigned int")
+      Res = ContextAST->UnsignedIntTy;
+    else if (Name == "long")
+      Res = ContextAST->LongTy;
+    else if (Name == "unsigned long")
+      Res = ContextAST->UnsignedLongTy;
+    else if (Name == "long long")
+      Res = ContextAST->LongLongTy;
+    else if (Name == "unsigned long long")
+      Res = ContextAST->UnsignedLongLongTy;
+    else if (Name == "__int128")
+      Res = ContextAST->Int128Ty;
+    else if (Name == "unsigned __int128")
+      Res = ContextAST->UnsignedInt128Ty;
+    else if (Name == "float")
+      Res = ContextAST->FloatTy;
+    else if (Name == "double")
+      Res = ContextAST->DoubleTy;
+    else if (Name == "long double")
+      Res = ContextAST->LongDoubleTy;
+    else if (Name == "__float128")
+      Res = ContextAST->Float128Ty;
+    else if (Name == "...") {
+      Res = clang::QualType{};
+      IsVariadic = true;
+    } else if (Name == "decimal64")
+      assert(false && "unhandled type name: decimal64");
+    else if (Name == "decimal128")
+      assert(false && "unhandled type name: decimal128");
+    else if (Name == "decimal32")
+      assert(false && "unhandled type name: decimal32");
+    else if (Name == "decimal16")
+      assert(false && "unhandled type name: decimal16");
+    else if (Name == "char32_t")
+      Res = ContextAST->Char32Ty;
+    else if (Name == "char16_t")
+      Res = ContextAST->Char16Ty;
+    else if (Name == "char8_t")
+      Res = ContextAST->Char8Ty;
+    else if (Name == "_Float16")
+      Res = ContextAST->Float16Ty;
+    else if (Name == "half")
+      Res = ContextAST->HalfTy;
+    else if (Name == "auto")
+      Res = ContextAST->AutoDeductTy;
+    else if (Name == "decltype(auto)")
+      assert(false && "unhandled type name: decltype(auto)");
+    else if (Name == "std::nullptr_t")
+      Res = ContextAST->NullPtrTy;
+    else {

```

Comment:
So if you don't know the type you assume it's an enumeration type? That seems dangerous. For example, this is missing a type for _BitInt, so that would turn into an enum here. I think the else case should be an assertion to help spot missing types.

Reply:
Although this matches what the previous code did.

Reply:
I've added an assert for `_BitInt`, the problem here is that we match on strings printed from `itanium_demangle::Node` and having a generic else close that matches an invalid state is not possible (or hard to get?). FWIW handling a valid qualified type as an enum (falling to catch all else as it is now) would result in a mis-mangled name and would be spotted by the tests.

Reply:
Doesn't that presume we're testing every possible builtin type? (That sounds like it would get out of sync pretty quickly as we do add new builtin types, though somewhat uncommonly.)

Reply:
This is similar to https://github.com/intel/llvm/pull/7220#discussion_r1028105906, we only need to be able to handle what spir-v can throw at the remangler.